### PR TITLE
Improve marshalling for flat points

### PR DESCRIPTION
**Pull request by furstenheim-geoblink**

### CODE REVIEW

File: coordseq.go [hunk 1]
```diff
@@ -50,26 +50,15 @@ func coordSeqFromSlice(coords []Coord) (*coordSeq, error) {
 func coordSeqFromFlatPoints(fp []float64) (*coordSeq, error) {
 	// XXX: handle 3-dim
 	length := len(fp)
-	if length & 1 == 1 {
+	if length&1 == 1 {
```
Comment (Sumolari) [inline coordseq.go:53]: Why this change?

File: coordseq.go [hunk 2]
```diff
@@ -50,26 +50,15 @@ func coordSeqFromSlice(coords []Coord) (*coordSeq, error) {
...
 		return nil, errors.New("Flat array should have even number of coordinates")
 	}
-	ptr := cGEOSCoordSeq_create(C.uint(length >> 1), C.uint(2))
+	ptr := cGEOSCoordSeq_create(C.uint(length>>1), C.uint(2))
```
Comment (Sumolari) [inline coordseq.go:56]: Why this change?

File: coordseq.go [hunk 3]
```diff
@@ -50,26 +50,15 @@ func coordSeqFromSlice(coords []Coord) (*coordSeq, error) {
...
 	if ptr == nil {
 		return nil, Error()
 	}
+	C.go_geos_FlatPointsToCoordSeq(handle, ptr, (*C.double)(&fp[0]), C.ulong(length>>1))
```
Comment (Sumolari) [inline coordseq.go:60]: Why not keeping the space between operands of `>>` operator?

File: geom.go [hunk 4]
```diff
@@ -415,6 +416,31 @@ func NewPolygonFromFlatPoints (shell []float64) (*Geometry, error){
...
+func (geom *Geometry) PolygonToFlatPoints(out []float64) ([]float64, error) {
+	shell, shellErr := geom.Shell()
+	if shellErr != nil {
+		return nil, shellErr
+	}
+	ptr := cGEOSGeom_getCoordSeq(shell.g)
+	if ptr == nil {
+		return nil, Error()
+	}
+	//cs := coordSeqFromPtr(ptr)
```
Comment (Sumolari) [inline geom.go]: Commented code

File: geom.go [hunk 5]
```diff
@@ -415,6 +416,31 @@ func NewPolygonFromFlatPoints (shell []float64) (*Geometry, error){
...
+	cs := &coordSeq{c: ptr}
+	nCoordinates, sizeErr := cs.size()
+	runtime.KeepAlive(cs)
+	if sizeErr != nil {
+		return nil, sizeErr
+	}
+	flatSize := nCoordinates * 2
```
Comment (Sumolari) [inline geom.go]: Why not using here the same trick you used before to multiply by 2? `>> 1`?

File: geos.c [hunk 6]
```diff
@@ -19,6 +19,28 @@ void go_geos_error_handler(const char *fmt, ...) {
     va_end(ap);
 }
 
+void go_geos_LinearRingToFlatPoints (GEOSContextHandle_t handler, double out[], size_t n_out, const GEOSCoordSequence *seq) {
+    int i;
+    double *j = out;
+    for (i = 0; i < n_out; i++) {
+        GEOSCoordSeq_getXY_r(handler, seq, i, j, j + 1);
+        j++;
```
Comment (Sumolari) [inline geos.c]: Why not `+= 2`? Using twice the same `++` operator looks like a mistake

File: geos.c [hunk 7]
```diff
@@ -19,6 +19,28 @@ void go_geos_error_handler(const char *fmt, ...) {
     va_end(ap);
 }
 
+void go_geos_LinearRingToFlatPoints (GEOSContextHandle_t handler, double out[], size_t n_out, const GEOSCoordSequence *seq) {
+    int i;
```
Comment (Sumolari) [inline geos.c:23]: Not a fan of having references to uninitialized memory regions, can you initialize this to `0` and then do nothing on the loop start?

**Review by Sumolari — APPROVED** [review]

File: coordseq.go (hunk 1 above)
Comment (furstenheim-geoblink) [reply coordseq.go:53]: This was automatic from the go fmt.

File: coordseq.go (hunk 2 above)
Comment (furstenheim-geoblink) [reply coordseq.go:56]: Same go fmt

File: geom.go (hunk 4 above)
Comment (furstenheim-geoblink) [reply geom.go]: In fact the original code was better since it set the finalizer

File: geom.go (hunk 5 above)
Comment (furstenheim-geoblink) [reply geom.go]: I suppose you mean << 1, fixed

File: geos.c (hunk 6 above)
Comment (furstenheim-geoblink) [reply geos.c]: Fixed

File: geos.c (hunk 7 above)
Comment (furstenheim-geoblink) [reply geos.c:23]: Fixed

**Comment by furstenheim-geoblink** [comment]
@Sumolari all comments addressed

**Review by zedfmario — APPROVED** [review]
👍
